xchange-stream-coinjar/src/main/java/info/bitrich/xchangestream/coinjar/CoinjarStreamingMarketDataService.java

```diff
@@ -0,0 +1,103 @@
+import info.bitrich.xchangestream.coinjar.dto.CoinjarWebSocketBookEvent;
+import info.bitrich.xchangestream.core.StreamingMarketDataService;
+import info.bitrich.xchangestream.service.netty.StreamingObjectMapperHelper;
+import io.reactivex.Observable;
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.Order;
+import org.knowm.xchange.dto.marketdata.OrderBook;
+import org.knowm.xchange.dto.marketdata.Ticker;
+import org.knowm.xchange.dto.marketdata.Trade;
+import org.knowm.xchange.dto.trade.LimitOrder;
+import org.knowm.xchange.exceptions.NotAvailableFromExchangeException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class CoinjarStreamingMarketDataService implements StreamingMarketDataService {
+
+  private static final Logger logger =
+      LoggerFactory.getLogger(CoinjarStreamingMarketDataService.class);
+
+  private final ObjectMapper mapper = StreamingObjectMapperHelper.getObjectMapper();
+
+  private final CoinjarStreamingService service;
+
+  private final Map<CurrencyPair, SortedMap<BigDecimal, LimitOrder>> bids = Maps.newConcurrentMap();
+  private final Map<CurrencyPair, SortedMap<BigDecimal, LimitOrder>> asks = Maps.newConcurrentMap();
+
+  public CoinjarStreamingMarketDataService(CoinjarStreamingService service) {
+    this.service = service;
+  }
+
+  private static void updateOrderbook(Map<BigDecimal, LimitOrder> book, List<LimitOrder> orders) {
+    orders.forEach(
+        order -> {
+          if (order.getOriginalAmount().compareTo(BigDecimal.ZERO) > 0) {
+            book.put(order.getLimitPrice(), order);
+          } else {
+            book.remove(order.getLimitPrice());
+          }
+        });
+  }
+
+  private OrderBook handleOrderbookEvent(CoinjarWebSocketBookEvent event) {
+    final CurrencyPair pairFromEvent =
+        CoinjarStreamingAdapters.adaptTopicToCurrencyPair(event.topic);
+    switch (event.event) {
+      case CoinjarWebSocketBookEvent.UPDATE:
+      case CoinjarWebSocketBookEvent.INIT:
+        updateOrderbook(
+            bids.get(pairFromEvent),
+            CoinjarStreamingAdapters.toLimitOrders(
+                event.payload.bids, pairFromEvent, Order.OrderType.BID));
+        updateOrderbook(
+            asks.get(pairFromEvent),
+            CoinjarStreamingAdapters.toLimitOrders(
+                event.payload.asks, pairFromEvent, Order.OrderType.ASK));
+        break;
+    }
+    return new OrderBook(
+        null,
+        Lists.newArrayList(asks.get(pairFromEvent).values()),
+        Lists.newArrayList(bids.get(pairFromEvent).values()));
+  }
+
+  @Override
+  public Observable<OrderBook> getOrderBook(CurrencyPair currencyPair, Object... args) {
+    String channelName = CoinjarStreamingAdapters.adaptCurrencyPairToBookTopic(currencyPair);
+    this.asks.put(currencyPair, Maps.newTreeMap(BigDecimal::compareTo));
```

Comment:
This doesn't look like it will play well with two concurrent subscriptions to the same currency pair.

Reply:
I've addressed the above issues and will look at this one, it's not a scenario i'm using but that should probably be fixed before merging.

Reply:
Having had another look at this, it's probably safe to roll out as is, each subscription has it's own orderbook which get updated for subscription. Although it's not the most efficient way when multiple subscriptions are present, that seems to be a common way of implementing this in various other streaming exchanges.